### DISCUSSION TITLE
MobX-React: Remove need for `observer` HOC, like Preact Signals & Legend State

### DISCUSSION THREAD

**Discussion by jamonholmgren** [mobxjs/mobx#3528] [Ideas]
Inspired by this tweet:

https://twitter.com/jmeistrich/status/1569644918464978945

It would be great to have a way to hook into React internals to remove the need for the `observer` HOC.

The way it seems to work is to hook into ReactCurrentDispatcher.current changing to detect the beginning/end of render, as described here: https://twitter.com/jmeistrich/status/1569644922382467072

Here's the relevant `enableLegendStateReact` function: https://github.com/LegendApp/legend-state/blob/65a4ae8681f098e9b529dd9790d420ee980ff26c/src/react/enableLegendStateReact.ts#L22

In React Native apps, if you export a component like this:

```tsx
export const ProfileAvatar = observer(() => {
  const { user } = useStores()
  
  return <Image src={user.profileURL} />
})
```

... it'll lose the name of the component, so debugging is difficult. To fix this, we've settled on this awkward syntax:

```tsx
export const ProfileAvatar = observer(function ProfileAvatar() {
  const { user } = useStores()
  
  return <Image src={user.profileURL} />
})
```

However, if we implemented the above idea, it would simply look like this:

```tsx
enableMobXReact()

export function ProfileAvatar() {
  const { user } = useStores()
  
  return <Image src={user.profileURL} />
}
```

Thoughts?

**Comment by kubk** [comment]
Initially I thought it's yet another "let's use hooks instead of observer" topic discussed here dozen of times, but then I realized it's something more interesting. Thanks for bringing this up! Unfortunately I am unable to provide qualified feedback here, so let's wait for other maintainers.

Am I understanding correctly that this patch is going to affect even vendor code (like Antd table or Material UI dropdown), so we won't need toJS for such cases anymore?

**Reply by jamonholmgren** [reply]
> Am I understanding correctly that this patch is going to affect even vendor code (like Antd table or Material UI dropdown), so we won't need toJS for such cases anymore?

I unfortunately haven't studied this implementation enough to know how it will affect them.

**Comment by iChenLei** [comment]
Interesting share : )

**Comment by mweststrate** [comment]
The primary concern I'd have here is that it would affect any React component (as far as I could tell from the code), including components from libraries etc. That sounds potentially quite dangerous leading to subtle bugs, as MobX introduces several assumptions, like applying `memo` automatically. So I think it'd be great to somehow still *somehow* mark the components. Could even be based on filepath or so, if that is still predictable after going through a build. But it'd be really annoying if a component in for example ant.design breaks because of MobX being used somewhere in the project.

**Reply by kubk** [reply]
Is there a way to avoid applying `memo` automatically in case the component is already memoized by a vendor library?

**Reply by jamonholmgren** [reply]
I understand theoretically, but I'm curious if this would actually cause any issues in reality.

How difficult would it be to make this a plugin that could be tested across many projects, but not part of core? Something like `mobx-react-observer` or something, and then it hooks up the right things under the hood. Only integrate into mobx-react-lite if it seems to work really well (or keep it separate). I think the benefits are great enough that it would be helpful to explore.

(And having to specify what components should be observers brings us back to decorator-land, which is ... ehhh)

**Reply by mweststrate** [reply]
I think exploring it is definitely a good idea! Publishing it as separate package with clear disclaimer could be very interesting for people to experiment with.

> (And having to specify what components should be observers brings us back to decorator-land, which is ... ehhh)

Haha yeah I know. It would primarily give some more flexibility then in how to achieve that. E.g. could also be based on component name, a `useMobX()` call, or whatever.

**Reply by mweststrate** [reply]
As interesting benefit, this might also avoid the need to slice / (shallow) clone objects before passing them to external components (I think someone else already mentioned that, but can't find the comment anymore)

**Comment by jmeistrich** [comment]
Author of Legend-State here with some technical details. Happy to share any more details if I can help.

The downside of the `enableLegendStateReact` behavior is that it requires injecting a useReducer and useEffect (I'm working on a way to remove the useEffect) into every single component (including vendor code) regardless of whether it tracks anything. If not, a component could track an observable conditionally and hit the "rendered fewer hooks" type of errors.

In practice I'm not sure what is better for performance - inject a useReducer into every component vs. the `observer` HOC on many components. I'd be very curious to benchmark it.

I don't think there's a way to detect file path in there. We can also track the ReactCurrentOwner, but it doesn't have much information beyond the props. If you want to do that I think you'd want to go a direction of making a babel plugin that auto-wraps components in `observer`?

But it leaves `memo` up to the user, so they might still end up wrapping every component in `memo` instead of `observe`.

If you're seriously looking into this, I'd suggest also looking at https://github.com/preactjs/signals/blob/main/packages/react/src/index.ts. That's what inspired my solution, but it has a different approach which also tracks the current owner fiber.

**Reply by jmeistrich** [reply]
For what it's worth, I benchmarked the dispatcher hooking method vs. the `observer` HOC and dispatcher hooking is generally a little bit faster. Of course it's a synthetic benchmark so it may not be totally accurate to real apps but it's a good data point, and it allays my worries of overhead caused by injecting `useReducer` everywhere.

**Comment by urugator** [comment]
@jamonholmgren

> However, if we implemented the above idea, it would simply look like this:

You would either loose `React.memo` applied by `observer` or you would have to apply it manually leading to the same problem. Same for any other HOC + anonymous fn. If the main concern is a component name a babel transform plugin would probably be a more realistic short term, production ready solution.
https://github.com/patrykkopycinski/babel-plugin-react-anonymous-display-name
https://github.com/laiff/babel-plugin-react-wrapped-display-name
https://github.com/zendesk/babel-plugin-react-displayname

**Reply by jamonholmgren** [reply]
Thanks @urugator, appreciate the feedback! The Babel plugin is useful for sure. I’d still like to explore this; how important is React.memo for observer function components?

**Reply by jmeistrich** [reply]
For what it's worth, using Legend-State I still wrap all components in `memo`. Well, actually a custom HOC that applies `memo` and an ErrorBoundary.

**Comment by k-ode** [comment]
Just as a side note, using anonymous hoc functions also silently breaks some eslint rules, like the React hooks one: https://github.com/facebook/react/issues/20499

For this reason we split all components into two:

```js
const ProfileAvatarComp = () => ...
export const ProfileAvatar = observer(ProfileAvatarComp);
```

Also remember the unobserved component working better with forwardRef and tools like Storybook.

**Comment by jmeistrich** [comment]
I've actually removed this behavior from Legend-State and I recommend you don't pursue it either. We found it to be extremely error prone, completely crashing apps with mismatched number of hooks errors. The errors are easy to reproduce in Next.js with just `router.replace(...)` in Next.js. I could not manage to figure out a way to solve all of the issues to make it stable. From the looks of the issues in the Signals repo they haven't either.

Also, the React team has asked me not to do it and said it is likely to change in a future version. So all in all it seems like a no-go.
